modules/storages/app/workers/manage_nextcloud_integration_job.rb

```diff
@@ -29,141 +29,192 @@
 class ManageNextcloudIntegrationJob < Cron::CronJob
   using ::Storages::Peripherals::ServiceResultRefinements
 
-  PERMISSION_MAP = {
+  PERMISSIONS_MAP = {
     read_files: 1,
     write_files: 2,
     create_files: 4,
     delete_files: 8,
     share_files: 16
   }.freeze
+  ALL_PERMISSIONS = PERMISSIONS_MAP.values.sum
+  NO_PERMISSIONS = 0
 
   queue_with_priority :low
 
   self.cron_expression = '*/5 * * * *'
 
   def perform
-    Storages::NextcloudStorage.where("provider_fields->>'has_managed_project_folders' = 'true'").each do |storage|
-      username = storage.username
-      group = storage.group
-      groupfolder = storage.groupfolder
-      nextcloud_usernames_used_in_openprojects = Set.new
-      project_folder_ids_used_in_openproject = Set.new
+    Storages::NextcloudStorage
+      .where("provider_fields->>'has_managed_project_folders' = 'true'")
+      .includes(:oauth_client)
+      .each do |storage|
+      NextcloudStorageManager.new(storage).call
+    end
+  end
 
-      requests = Storages::Peripherals::StorageRequests.new(storage:)
+  class NextcloudStorageManager
```

Comment:
I really like that the main logic is now extracted and could be used somewhere else. I'm not fine with the name though. A `...Manager` implies a longer living object that gets hold of something that needs to be managed. This is not the case here, as this class only contain extracted methods. Basically it's still one function that does a specific thing: Fixing and applying settings to the group folder and users/groups in the storage.

I don't want to inflate this here even more, but later we can extract all single methods as `public` method to a utility class for changing things in the storage with regards to project folders. Than have a small `ProjectFolderRepairer` class that make use of the methods from the utility class. This is then executed by the `Job`. What do you think. It would be nice to do this for the single trigger points.

Reply:
> I really like that the main logic is now extracted and could be used somewhere else. I'm not fine with the name though. A `...Manager` implies a longer living object that gets hold of something that needs to be managed. This is not the case here, as this class only contain extracted methods. Basically it's still one function that does a specific thing: Fixing and applying settings to the group folder and users/groups in the storage

I do not like the name as well. I did not find a better one and decided to stick with it for a time being. Maybe, `NextcloudStorageSynchonizer` or `NetxcouldStorageRepairer` would work. Wdyt?

> I don't want to inflate this here even more, but later we can extract all single methods as `public` method to a utility class for changing things in the storage with regards to project folders. Than have a small `ProjectFolderRepairer` class that make use of the methods from the utility class. This is then executed by the `Job`. What do you think. It would be nice to do this for the single trigger points.

I am not sure that I completely get the idea. I would stick to a single big class/module and have things in one place till the moment we really "fix" things sparingly as a reaction on a particular event in OP (user has been added to a project => we only trigger appropriate updates in NC).

Reply:
> I am not sure that I completely get the idea. I would stick to a single big class/module and have things in one place till the moment we really "fix" things sparingly as a reaction on a particular event in OP (user has been added to a project => we only trigger appropriate updates in NC).

That's what I meant. Not now, but later for the events. Good.

> I do not like the name as well. I did not find a better one and decided to stick with it for a time being. Maybe, `NextcloudStorageSynchonizer` or `NetxcouldStorageRepairer` would work. Wdyt?

I would go with something like `ProjectFolderPropertiesSynchronization`. A bit long but maybe there is a shorter one.

Reply:
> That's what I meant. Not now, but later for the events. Good.

:+1: 


> I would go with something like `ProjectFolderPropertiesSynchronization`. A bit long but maybe there is a shorter one.

I would expect `ProjectFolderPropertiesSynchronization` to operate on one project folder and set its properties, but current class operates in scope of `group_folder` and its one level children:
1. sets permissions for them
2. create/rename/hide project_folders
3. Add/remove users to/from `group`

What about the following signature: `GroupFolderSynchronization.new(storage).call`?

Reply:
> What about the following signature: `GroupFolderSynchronization.new(storage).call`?

I would assume that it's syncing files in a group folder.

I would like to keep the "Properties" part. And to that, "group folders" are only one way of implementing the solution. And only for next cloud. The class should operate independently from the type of storage. Therefor it is Nextcloud only at the moment.

But we should get done now. `GroupFolderPropertiesSynchronization`?

Reply:
Agreed, let it be `GroupFolderPropertiesSynchronization`